src/VecchiaMLE_kernels.jl

```diff
@@ -106,3 +127,89 @@ function vecchia_build_B!(B::Vector{Matrix{T}}, samples::Matrix{T}, rowsL::Vecto
     end
     return nothing
 end
+
+#=
+    VECCHIA_GENERATE_HESS_TRI_STRUCTURE
+    Front-end
+    kernel
+    CPU implementation
+=#
+
+# Front-end
+function vecchia_generate_hess_tri_structure!(nnzh::Int, n::Int, colptr_diff::CuVector{Int}, 
+    hrows::CuVector{Int}, hcols::CuVector{Int})
+
+    # reset hrows, hcols
+    fill!(hrows, zero(Int))
+    fill!(hcols, zero(Int))
+
+    # launch the kernel
+    backend = KA.get_backend(hrows)
+    kernel = vecchia_generate_hess_tri_structure_kernel!(backend)
+
+    f(x) = (x * (x+1)) ÷ 2
+
+    carry_offsets = CUDA.ones(Int, n)
+    view(carry_offsets, 2:n) .+= cumsum(f.(view(colptr_diff, 1:n-1)))
+
+    idx_offsets = CUDA.ones(Int, n)
+    view(idx_offsets, 2:n) .+= cumsum(view(colptr_diff, 1:n-1))
+
+    kernel(nnzh, n, colptr_diff, carry_offsets, idx_offsets, hrows, hcols, ndrange = n)
+
+    KA.synchronize(backend)
+
+    return nothing
+end
+
+# kernel
+@kernel function vecchia_generate_hess_tri_structure_kernel!(
+    @Const(nnzh), @Const(n), @Const(colptr_diff), @Const(carry_offsets), @Const(idx_offsets),
+    hrows, hcols
+)
+    thread_idx = @index(Global) # in 1:n
+    m = colptr_diff[thread_idx]
+    carry = carry_offsets[thread_idx]
+    idx = idx_offsets[thread_idx]
+
+    for j in 1:m
+        for k in carry:m - j + carry
+            hrows[k] = (idx + j - 1) + (k - carry)
+            hcols[k] = idx + j - 1
+        end
+        carry += m - j + 1
+    end
+
+    # fill one index of the tail for each thread
+    @inbounds hrows[nnzh-n + thread_idx] = hrows[nnzh-n] + thread_idx
+    @inbounds hcols[nnzh-n + thread_idx] = hrows[nnzh-n] + thread_idx
+end
+
+
+
+# CPU implementation
+function vecchia_generate_hess_tri_structure!(nnzh::Int, n::Int, colptr_diff::Vector{Int}, 
+    hrows::Vector{Int}, hcols::Vector{Int}) 
+    
+    carry = 1
+    idx = 1
+    for i in 1:n
+        m = colptr_diff[i]
+            for j in 1:m
+                view(hrows, (0:(m-j)).+carry) .= (j:m).+(idx-1)
+                fill!(view(hcols, carry:carry+m-j), idx + j - 1)
+                carry += m - j + 1
+            end
+        idx += m
+    end
+
+    #Then need the diagonal tail
+    idx_to = idx + nnzh - carry
+    view(hrows, carry:nnzh) .= idx:idx_to
+    view(hcols, carry:nnzh) .= idx:idx_to
+
+    println("CPU:\n")
+    println(hrows)
+    println(hcols)
```

Comment:
@CalebDerrickson You can't keep these debug printing when you merge your modifcations into the main branch.